### ISSUE TITLE
Span should support async context manager protocol

### ISSUE THREAD

**Issue by notypecheck** [getsentry/sentry-python#2007]
### Problem Statement

`Span` and `Transaction` should support async context manager protocol to avoid unnecessary nesting.

As an example:
```py
async def main() -> None:
    async with a:
        with sentry_sdk.start_transaction(
            op="op",
            name="name",
        ):
            with sentry_sdk.start_span(op="b"):
                async with b:
                    await b.run()

            with sentry_sdk.start_span(op="c"):
                async with c:
                    await c.run()
```

Could be shortened to
```py
async def main() -> None:
    async with a, sentry_sdk.start_transaction(
        op="op",
        name="name",
    ):
        async with sentry_sdk.start_span(op="b"), b:
            await b.run()

        async with sentry_sdk.start_span(op="c"), c:
            await c.run()
```

### Solution Brainstorm

I think currently `__aenter__` and `__aexit__` could do the same as their synchronous counterparts and later could be replaced with async code, if that would be possible at some point 🤔

**Comment by antonpirker** [comment]
Hey @ThirVondukr !

Thanks for bringing this up. Yes, our async Python support needs some improvements. 
I put this on our backlog, but we are quite swamped right now, so it will take some time for us to implement this.

But pull requests are always welcomed, if you (or anyone else) would like to give it a try!

**Comment by ThirVondukr** [comment]
@antonpirker I think to make these objects compatible with async (just to reduce nesting for now) they could call their sync counterparts, would that implementation suffice? 🤔

**Comment by antonpirker** [comment]
Probably! If you submit a PR, I am happy to review!

**Comment by isra17** [comment]
This is a pretty bad idea as per https://github.com/getsentry/sentry-python/issues/1735 .

You would pretty much be lying through the API signature that sentry is async-compatible while it’s clearly not. As it turns out, all the async framework integration is very bad as well. Any error will stop the whole event loop until the event has been captured and sent to the sentry server. This is a big no-no for async code.

In my opinion, the correct thing right now would be to put a big warning about how sentry integration are not working as intended for async codes and framework and that big performance penalty will come with their use.

This might come out a bit harsh, but I just come out of a performance debug session about a long standing issue and I was quite surprised that only one person raised this flag in the issues. Doing async API is not as easy as slapping the `async` keyword in front of any function.

**Comment by ThirVondukr** [comment]
@isra17 This could be an issue, but it depends on your workload, since event is sent only when you exit the transaction, so it's not a big issue for low RPS operations/processes, like ETL
Maybe it would be possible to use something like greenlet here?

**Comment by isra17** [comment]
Actually it seems like the default transport use a background worker to send events, I still need to investigate why it seems to block the event loop 🤔

**Comment by sl0thentr0py** [comment]
yea the 'heavy' (sending over network) work shouldn't be blocking since that happens in the background worker thread, but the serialization happens in the original flow.

**Comment by isra17** [comment]
Yeah, it seems like my loop being blocked in my repro script was due to `sentry.init`. I was wrong, sorry about that!
